### PR TITLE
workflow training neural network from warpx data

### PR DESCRIPTION
**Pull request by RTSandberg**
Document a sample machine learning workflow with data from a WarpX simulation.  This workflow includes the following steps:
* cleaning data
    * Because some bunches look good (the training beam for stage 1)
stage_1_final_beam
    * But some beams do not (the training beam for stage 0)
stage_0_final_beam
* formatting data 
* training a neural network
* evaluating the neural network
    * Which here means plotting the train/test loss curves:
beam_stage_0_training_testing_error
    * and projections of the final phase space
beam_stage_0_model_evaluation

### CODE REVIEW

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 1]
```diff
@@ -0,0 +1,254 @@
+.. _ml_dataset_training:
+
+Training a surrogate model from WarpX data
+==========================================
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
Training a Surrogate Model from WarpX Data
==========================================
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+.. dropdown:: Python Input for Training Simulation
+    :color: light
+    :icon: info
+    :animate: fade-in-slide-down
+
+    .. literalinclude:: ml_materials/run_warpx_training.py
+    :language: python3
+
+In this section we walk through a workflow for data processing and model training.
+This workflow was developed and first presented in (CITE PASC submission).
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]: todo :)
```suggestion
This workflow was developed and first presented in Refs. :cite:t:`SandbergIPAC23,SandbergPASC24`.
```
and add in our bib file for now:
```bib
@misc{SandbergPASC24,
    author = {Ryan Sandberg and Remi Lehe and Chad E Mitchell and Marco Garten and Ji Qiang and Jean-Luc Vay and Axel Huebl},
    title = {{Synthesizing Particle-in-Cell Simulations Through Learning and GPU Computing for Hybrid Particle Accelerator Beamlines}},
    booktitle = {Proc. of PASC24},
    venue = {Zuerich, Switzerland},
    address = {Zuerich, Switzerland},
    series = {PASC'24 - Platform for Advanced Scientific Computing},
    year = {2024},
    note = {submitted}
}

@inproceedings{SandbergIPAC23,
    author = {Ryan Sandberg and Remi Lehe and Chad E Mitchell and Marco Garten and Ji Qiang and Jean-Luc Vay and Axel Huebl},
    title = {{Hybrid beamline element ML-training for surrogates in the ImpactX beam-dynamics code}},
    booktitle = {Proc. 14th International Particle Accelerator Conference},
    pages = {2885-2888},
    paper = {WEPA101},
    venue = {Venice, Italy},
    address = {Venice, Italy},
    series = {IPAC'23 - 14th International Particle Accelerator Conference},
    number = {14},
    publisher = {JACoW Publishing, Geneva, Switzerland},
    month = {05},
    year = {2023},
    issn = {2673-5490},
    isbn = {978-3-95450-231-8},
    doi = {10.18429/JACoW-IPAC2023-WEPA101},
    url = {https://indico.jacow.org/event/41/contributions/2276},
    language = {English}
}
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+.. figure:: ml_materials/stage_1_final_beam.png
+   :alt: [fig:stage_1_final_beam] Plot comparing model prediction with simulation output.
+
+   [fig:stage_1_final_beam] Plot showing the final phase space projections of the training data for stage 1.
+
+This is as we expect and what is optimal for training neural networks.
+On the other hand, the final phase space for beam 0 has a halo of outlying particles.
+
+.. figure:: ml_materials/stage_0_final_beam.png
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]: Please add the PNGs in the PR description and use the link to them (do not add to `git` to keep the WarpX repo small :) )

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+.. dropdown:: Python neural network training
+    :color: light
+    :icon: info
+    :animate: fade-in-slide-down
+
+    .. literalinclude:: ml_materials/train.py
+    :language: python3
+
+Training function
+^^^^^^^^^^^^^^
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
Training Function
^^^^^^^^^^^^^^^^^
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+Note that this function returns the sum of all errors across the entire dataset,
+which is later divided by the size of the dataset in the training loop.
+
+.. literalinclude:: ml_materials/train.py
+   :language: python3
+   :start-after: # Manual: Train function START
+   :end-before: # Manual: Train function END
+
+Tesing function
+^^^^^^^^^^^^^
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
Testing Function
^^^^^^^^^^^^^^^^
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+First we consider the train-test loss curves, shown in Fig. `[fig:train_test_loss] <#fig:train-test>`__ .
+This figure shows the model error on the training data (in blue) and testing data (in green) as a function of the number of epochs seen.
+The training data is used to update the model parameters, so training error should be lower than testing error.
+A key feature to look for in the train-test loss curve is the inflection point in the test loss trend.
+The testing data is set aside as a sample of data the neural network hasn't seen before.
+The testing error serves as a metric of model generalizability, indicating how well the model performs
+on data it hasn't seen yet.
+When the test-loss starts to trend flat or even upward, the neural network is no longer improving its ability to generalize to new data.
+
+.. figure:: ml_materials/beam_stage_0_training_testing_error.png
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]: See above :)

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+   :alt: [fig:train_test_loss] Plot of training and testing loss curves
+
+   [fig:train_test_loss] Plot of training (in blue) and testing (in green) loss curves versus number of training epochs.
+
+A visual inspection of the model prediction can be seen in Fig. `[fig:train_evaluation]` .
+This plot compares the model prediction, with dots colored by mean-square error, on the testing data with the actual simulation output in black.
+
+.. figure:: ml_materials/beam_stage_0_model_evaluation.png
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]: See above :)

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+    :icon: info
+    :animate: fade-in-slide-down
+
+    .. literalinclude:: ml_materials/visualize.py
+    :language: python3
+
+Reference ImpactX documentation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+A neural network such as the one we trained here can be incorporated in other BLAST codes.
+Consider the example here (link to ImpactX) using neural networks in ImpactX.
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]: Link will be after is merged:
```suggestion
`Consider the example using neural networks in ImpactX <https://impactx.readthedocs.io/en/latest/usage/examples/pytorch_surrogate_model/README.html>`__.
```

**Review by ax3l — COMMENTED** [review]
Very nice docs! Added review comments inline :)

File: Docs/source/usage/workflows/ml_materials/create_dataset.py [hunk 9]
```diff
@@ -0,0 +1,192 @@
+
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_materials/create_dataset.py]:
```suggestion
#!/usr/bin/env python3
#
# Copyright 2023 The WarpX Community
#
# This file is part of WarpX.
#
# Authors: Ryan Sandberg
# License: BSD-3-Clause-LBNL
#
```

File: Docs/source/usage/workflows/ml_materials/neural_network_classes.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
+
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_materials/neural_network_classes.py]:
```suggestion
#!/usr/bin/env python3
#
# Copyright 2023 The WarpX Community
#
# This file is part of WarpX.
#
# Authors: Ryan Sandberg
# License: BSD-3-Clause-LBNL
#
```

File: Docs/source/usage/workflows/ml_materials/run_warpx_training.py [hunk 11]
```diff
@@ -0,0 +1,347 @@
+#!/usr/bin/env python3
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_materials/run_warpx_training.py:1]:
```suggestion
#!/usr/bin/env python3
#
# Copyright 2023 The WarpX Community
#
# This file is part of WarpX.
#
# Authors: Ryan Sandberg
# License: BSD-3-Clause-LBNL
#
```

File: Docs/source/usage/workflows/ml_materials/train.py [hunk 12]
```diff
@@ -0,0 +1,159 @@
+
+
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_materials/train.py]:
```suggestion
#!/usr/bin/env python3
#
# Copyright 2023 The WarpX Community
#
# This file is part of WarpX.
#
# Authors: Ryan Sandberg
# License: BSD-3-Clause-LBNL
#
```

File: Docs/source/usage/workflows/ml_materials/visualize.py [hunk 13]
```diff
@@ -0,0 +1,157 @@
+
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_materials/visualize.py]:
```suggestion
#!/usr/bin/env python3
#
# Copyright 2023 The WarpX Community
#
# This file is part of WarpX.
#
# Authors: Ryan Sandberg
# License: BSD-3-Clause-LBNL
#
```

File: Docs/source/usage/workflows/ml_dataset_training.rst (hunk 6 above)
Comment (RTSandberg) [reply Docs/source/usage/workflows/ml_dataset_training.rst]: done

File: Docs/source/usage/workflows/ml_dataset_training.rst (hunk 7 above)
Comment (RTSandberg) [reply Docs/source/usage/workflows/ml_dataset_training.rst]: done

File: Docs/source/usage/workflows/ml_materials/neural_network_classes.py [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+from enum import Enum
+
+import torch
```
[the remark this answers is not in the document]
Comment (RTSandberg) [reply Docs/source/usage/workflows/ml_materials/neural_network_classes.py]: acknowledged andfixed

**Review by RTSandberg — COMMENTED** [review]
addressing review comments

File: Docs/source/usage/workflows/ml_dataset_training.rst (hunk 3 above)
Comment (RTSandberg) [reply Docs/source/usage/workflows/ml_dataset_training.rst]: done!

**Review by ax3l — APPROVED** [review]
Let's rewrite the download logic another time, since we currently control the source this is not a big issue right now.

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+Suppose we have a WarpX simulation that we wish to replace with a neural network surrogate model.
+For example, a simulation determined by the following input script
+
+.. dropdown:: Python Input for Training Simulation
+    :color: light
+    :icon: info
+    :animate: fade-in-slide-down
+
+    .. literalinclude:: ml_materials/run_warpx_training.py
+    :language: python3
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
.. dropdown:: Python Input for Training Simulation
   :color: light
   :icon: info
   :animate: fade-in-slide-down

    .. literalinclude:: ml_materials/run_warpx_training.py
       :language: python
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+
+It is important to inspect the data for artifacts to
+check that input/output data make sense.
+If we plot the final phase space for beams 1-8,
+the particle data is distributed in a single blob.
+
+.. figure:: https://user-images.githubusercontent.com/10621396/290010209-c55baf1c-dd98-4d56-a675-ad3729481eee.png
+   :alt: [fig:stage_1_final_beam] Plot comparing model prediction with simulation output.
+
+   [fig:stage_1_final_beam] Plot showing the final phase space projections of the training data for stage 1.
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
   :alt: Plot comparing model prediction with simulation output.

   Plot showing the final phase space projections of the training data for stage 1.
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 17]
```diff
@@ -0,0 +1,254 @@
...
+
+This is as we expect and what is optimal for training neural networks.
+On the other hand, the final phase space for beam 0 has a halo of outlying particles.
+
+.. figure:: https://user-images.githubusercontent.com/10621396/290010282-40560ac4-8509-4599-82ca-167bb1739cff.png
+   :alt: [fig:stage_0_final_beam] Plot comparing model prediction with simulation output.
+
+   [fig:stage_0_final_beam] Plot showing the final phase space projections of the training data for stage 1.
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
   :alt: Plot comparing model prediction with simulation output.

   Plot showing the final phase space projections of the training data for stage 1.
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 18]
```diff
@@ -0,0 +1,254 @@
...
+
+Looking closer at the z-pz space, we see that some particles got caught in a decelerating
+region of the wake, have slipped back and are much slower than the rest of the beam.
+To assist our neural network in learning dynamics of interest, we filter out these particles.
+It is sufficient for our purposes to select particles that are not too far back, setting
+``particle_selection={'z':[0.28002, None]}``. Then a particle tracker is set up to make sure
+we consistently filter out these particles from both the initial and final data.
+
+.. literalinclude:: ml_materials/create_dataset.py
+   :language: python3
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
   :language: python
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 19]
```diff
@@ -0,0 +1,254 @@
...
+Load openPMD Data
+^^^^^^^^^^^^^^^^^
+
+First the openPMD data is loaded, using the particle selector as chosen above.
+The neural network will make predictions from the initial phase space coordinates,
+using the final phase space coordinates to measure how well it is making predictions.
+Hence we load two sets of particle data, the source and target particle arrays.
+
+.. literalinclude:: ml_materials/create_dataset.py
+   :language: python3
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
.. literalinclude:: ml_materials/create_dataset.py
   :language: python
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+The script below will take the openPMD data we have selected and
+format, normalize, and store it.
+
+.. dropdown:: Python dataset creation
+    :color: light
+    :icon: info
+    :animate: fade-in-slide-down
+
+    .. literalinclude:: ml_materials/create_dataset.py
+    :language: python3
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
.. dropdown:: Python dataset creation
   :color: light
   :icon: info
   :animate: fade-in-slide-down

    .. literalinclude:: ml_materials/create_dataset.py
       :language: python
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+Normalize Data
+^^^^^^^^^^^^^^
+
+Neural networks learn better on appropriately normalized data.
+Here we subtract out the mean in each coordinate direction and
+divide by the standard deviation in each coordinate direction,
+for normalized data that is centered on the origin with unit variance.
+
+.. literalinclude:: ml_materials/create_dataset.py
+   :language: python3
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
   :language: python
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 22]
```diff
@@ -0,0 +1,254 @@
...
+openPMD to PyTorch Data
+^^^^^^^^^^^^^^^^^^^^^^^
+
+With the data normalized, it must be stored in a form PyTorch recognizes.
+The openPMD data are 6 lists of arrays, for each of the 6 phase space coordinates
+:math:`x, y, z, p_x, p_y,` and :math:`p_z`.
+This data are converted to an :math:`N\times 6` numpy array and then to a PyTorch :math:`N\times 6` tensor.
+
+.. literalinclude:: ml_materials/create_dataset.py
+   :language: python3
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
   :language: python
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 23]
```diff
@@ -0,0 +1,254 @@
...
+   :end-before: # Manual: Format data END
+
+Save Normalizations and Normalized Data
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+With the data properly normalized, it and the normalizations are saved to file for
+use in training and inference.
+
+.. literalinclude:: ml_materials/create_dataset.py
+   :language: python3
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
   :language: python
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 24]
```diff
@@ -0,0 +1,254 @@
...
+The script below trains the neural network on the dataset just created.
+In subsequent sections we discuss the various parts of the training process.
+
+.. dropdown:: Python neural network training
+    :color: light
+    :icon: info
+    :animate: fade-in-slide-down
+
+    .. literalinclude:: ml_materials/train.py
+    :language: python3
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
.. dropdown:: Python neural network training
   :color: light
   :icon: info
   :animate: fade-in-slide-down

    .. literalinclude:: ml_materials/train.py
       :language: python3
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 25]
```diff
@@ -0,0 +1,254 @@
...
+is determined when the loss function is evaluated and the ``loss.backward()`` function
+is invoked.
+The optimizer uses this information to update the weights in the ``optimizer.step()`` call.
+The training loop then resets the optimizer and updates the summed error for the whole dataset
+with the error on the batch and continues iterating through batches.
+Note that this function returns the sum of all errors across the entire dataset,
+which is later divided by the size of the dataset in the training loop.
+
+.. literalinclude:: ml_materials/train.py
+   :language: python3
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
   :language: python
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 26]
```diff
@@ -0,0 +1,254 @@
...
+The testing function just evaluates the neural network on the testing data that has not been used
+to update the model parameters.
+This testing function requires that the testing dataset is small enough to be loaded all at once.
+The PyTorch dataloader can load data in batches if this size assumption is not satisfied.
+The error, measured by the loss function, is returned by the testing function to be aggregated and stored.
+Note that this function returns the sum of all errors across the entire dataset,
+which is later divided by the size of the dataset in the training loop.
+
+.. literalinclude:: ml_materials/train.py
+   :language: python3
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
   :language: python
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 27]
```diff
@@ -0,0 +1,254 @@
...
+Train Loop
+^^^^^^^^^^
+
+The full training loop performs ``n_epochs`` number of iterations.
+At each iteration the training and testing functions are called,
+the respective errors are divided by the size of the dataset and recorded,
+and a status update is printed to the console.
+
+.. literalinclude:: ml_materials/train.py
+   :language: python3
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
   :language: python
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 28]
```diff
@@ -0,0 +1,254 @@
...
+
+Save Neural Network Parameters
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The model weights are saved after training to record the updates to the model parameters.
+Addtionally, we save some model metainformation with the model for convenience,
+including the model hyperparameters, the training and testing losses, and how long the training took.
+
+.. literalinclude:: ml_materials/train.py
+   :language: python3
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
   :language: python
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 29]
```diff
@@ -0,0 +1,254 @@
...
+A key feature to look for in the train-test loss curve is the inflection point in the test loss trend.
+The testing data is set aside as a sample of data the neural network hasn't seen before.
+The testing error serves as a metric of model generalizability, indicating how well the model performs
+on data it hasn't seen yet.
+When the test-loss starts to trend flat or even upward, the neural network is no longer improving its ability to generalize to new data.
+
+.. figure:: https://user-images.githubusercontent.com/10621396/290010428-f83725ab-a08f-494c-b075-314b0d26cb9a.png
+   :alt: [fig:train_test_loss] Plot of training and testing loss curves
+
+   [fig:train_test_loss] Plot of training (in blue) and testing (in green) loss curves versus number of training epochs.
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
   :alt: Plot of training and testing loss curves

   Plot of training (in blue) and testing (in green) loss curves versus number of training epochs.
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 30]
```diff
@@ -0,0 +1,254 @@
...
+
+A visual inspection of the model prediction can be seen in Fig. `[fig:train_evaluation]` .
+This plot compares the model prediction, with dots colored by mean-square error, on the testing data with the actual simulation output in black.
+
+.. figure:: https://user-images.githubusercontent.com/10621396/290010486-4a3541e7-e0be-4cf1-b33b-57d5e5985196.png
+   :alt: [fig:train_evaluation] Plot comparing model prediction with simulation output.
+
+   [fig:train_evaluation] Plot comparing model prediction (yellow-red dots, colored by mean-squared error) with simulation output (black dots).
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
   :alt: Plot comparing model prediction with simulation output.

   Plot comparing model prediction (yellow-red dots, colored by mean-squared error) with simulation output (black dots).
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 31]
```diff
@@ -0,0 +1,254 @@
...
+
+These figures can be generated with the following Python script.
+
+.. dropdown:: Python visualization of progress training neural network
+    :color: light
+    :icon: info
+    :animate: fade-in-slide-down
+
+    .. literalinclude:: ml_materials/visualize.py
+    :language: python3
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
.. dropdown:: Python visualization of progress training neural network
   :color: light
   :icon: info
   :animate: fade-in-slide-down

    .. literalinclude:: ml_materials/visualize.py
       :language: python3
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 32]
```diff
@@ -0,0 +1,254 @@
...
+.. dropdown:: Python Input for Training Simulation
+   :color: light
+   :icon: info
+   :animate: fade-in-slide-down
+
+    .. literalinclude:: ml_materials/run_warpx_training.py
+       :language: python
+
+In this section we walk through a workflow for data processing and model training.
+This workflow was developed and first presented in Refs. :cite:t:`SandbergIPAC23,SandbergPASC24`.
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
This workflow was developed and first presented in Refs. :cite:p:`SandbergIPAC23,SandbergPASC24`.
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 33]
```diff
@@ -0,0 +1,254 @@
...
+.. dropdown:: Python Input for Training Simulation
+   :color: light
+   :icon: info
+   :animate: fade-in-slide-down
+
+    .. literalinclude:: ml_materials/run_warpx_training.py
+       :language: python
+
+In this section we walk through a workflow for data processing and model training.
+This workflow was developed and first presented in Refs. :cite:p:`SandbergIPAC23,SandbergPASC24`.
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
This workflow was developed and first presented in Refs. :cite:t:`SandbergIPAC23` and :cite:t:`SandbergPASC24`.
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 34]
```diff
@@ -0,0 +1,254 @@
...
+.. literalinclude:: ml_materials/create_dataset.py
+   :language: python
+   :dedent: 4
+   :start-after: # Manual: Save dataset START
+   :end-before: # Manual: Save dataset END
+
+Neural Network Structure
+------------------------
+
+It was found in (cite PASC) that reasonable surrogate models are obtained with
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
It was found in Ref. :cite:p:`SandbergPASC24` that reasonable surrogate models are obtained with
```

**Review by ax3l — APPROVED** [review]

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 35]
```diff
@@ -0,0 +1,254 @@
...
+This plot compares the model prediction, with dots colored by mean-square error, on the testing data with the actual simulation output in black.
+
+.. figure:: https://user-images.githubusercontent.com/10621396/290010486-4a3541e7-e0be-4cf1-b33b-57d5e5985196.png
+   :alt: Plot comparing model prediction with simulation output.
+
+   Plot comparing model prediction (yellow-red dots, colored by mean-squared error) with simulation output (black dots).
+
+The model obtained with the hyperparameters chosen here trains quickly but is not very accurate.
+A more accurate model is obtained with 5 hidden layers and 800 nodes per layer,
+as discussed in (reference PASC).
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion
as discussed in :cite:t:`SandbergPASC24`.
```

File: Docs/source/usage/workflows/ml_dataset_training.rst [hunk 36]
```diff
@@ -0,0 +1,254 @@
...
+.. dropdown:: Python visualization of progress training neural network
+   :color: light
+   :icon: info
+   :animate: fade-in-slide-down
+
+    .. literalinclude:: ml_materials/visualize.py
+       :language: python3
+
+Reference ImpactX Documentation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (ax3l) [inline Docs/source/usage/workflows/ml_dataset_training.rst]:
```suggestion

Surrogate Usage in Accelerator Physics
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

```

**Pull request merged**